src/menhir.ml

```diff
@@ -102,31 +145,114 @@ module Run (P : PARAMS) = struct
+
+  (* [process3 stanza] converts a Menhir stanza into a set of build rules.
+     This is the three-step process where Menhir is invoked twice and OCaml
+     type inference is performed in between. *)
+
+  let process3 base (stanza : stanza) : unit =
+
+    (* 1. A first invocation of Menhir creates a mock [.ml] file. *)
+
+    rule_menhir
+      [ As stanza.flags
+      ; Deps (sources stanza.modules)
+      ; As [ "--base" ; base ]
+      ; A "--infer-write-query"; Target (Path.relative dir (mock_ml base))
+      ];
 
-     [Dyn (fun flags -> As flags)]
-                      indicates that any flags that appear in the stanza
-                      must be transmitted to Menhir.
+    (* 2. The OCaml compiler performs type inference. *)
 
-   *)
+    let name = Module.Name.of_string (mock base) in
 
-  let process (stanza : stanza) =
-    let base : string = Option.value_exn stanza.merge_into in
-    let args : args =
-      [ Dyn (fun flags -> As flags)
+    let mock_module : Module.t =
+      Module.make
+        name
+        ~impl:{ name = mock_ml base; syntax = OCaml }
+    in
+
+    (* The following incantation allows the mock [.ml] file to be preprocessed
+       by the user-specified [ppx] rewriters. *)
+
+    let mock_module =
+      Preprocessing.pp_module_as
+        (Compilation_context.preprocessing cctx)
+        name
+        mock_module
+        ~lint:false
+    in
+
+    Module_compilation.ocamlc_i
+      ~dep_graphs:(Ocamldep.rules_for_auxiliary_module cctx mock_module)
+      cctx
+      mock_module
+      ~output:(inferred_mli base);
+
+    (* 3. A second invocation of Menhir reads the inferred [.mli] file. *)
+
+    rule_menhir
+      [ As stanza.flags
       ; Deps (sources stanza.modules)
       ; As [ "--base" ; base ]
+      ; A "--infer-read-reply"; Dep (inferred_mli base)
       ; Hidden_targets (targets base)
       ]
+
+  (* ------------------------------------------------------------------------ *)
+
+  (* [process3 stanza] converts a Menhir stanza into a set of build rules.
+     This is a simpler one-step process where Menhir is invoked directly. *)
+
+  let process1 base (stanza : stanza) : unit =
+    rule_menhir
+      [ As stanza.flags
+      ; Deps (sources stanza.modules)
+      ; As [ "--base" ; base ]
+      ; Hidden_targets (targets base)
+      ]
+
+  (* ------------------------------------------------------------------------ *)
+
+  (* [process stanza] converts a Menhir stanza into a set of build rules, using
+     either [process3] or [process1], as appropriate. *)
+
+  (* Because Menhir processes [--only-tokens] before the [--infer-*] commands,
+     when [--only-tokens] is present, no [--infer-*] command should be used. *)
+
+  let process (stanza : stanza) : unit =
+    let base = Option.value_exn stanza.merge_into in
+    let ocaml_type_inference_disabled =
+      List.mem "--only-tokens" ~set:stanza.flags
     in
-    rule (menhir args)
+    if ocaml_type_inference_disabled then
```

Comment:
It's a bit sad to lose the ability to use `Ordered_set_lang.t` because of this. Would it make sense for this option to be toggled with a separate field in the stanza if there's no workaround?

The reason why it's nice to have `Ordered_set_lang.t` btw is that it lets you change the set of flags in a future proof way. Eventually, we'll have the ability to change what `:standard` will mean for a project and it would be nice if such changes propagated.

Reply:
I could revert to `Ordered_set_lang.t` if you wish.
Then, how would I implement your other suggestion of giving a precise error location when a reserved flag is given by the user? It seems that `Super_context.expand_and_eval_set` returns an action of type `(unit, string list) Build.t` which does not give me access to the location of each individual flag.

Reply:
I suggest the following: we add a function in `Ordered_set_lang.Unexpanded.t` to scan the the static strings. Then in menhir we do the following:

- scan the flags for a plain `--only-tokens` flag and select the behaviour accordingly
- if `--only-tokens` wasn't present, scan the flags after they are fully evaluated and error out if `--only-tokens` is present

Reply:
That sounds good. It should still allow catching the cases where people pass incorrect static flags, and would still allow using your expansion mechanism inside flags.

Reply:
Okay, this behavior seems good to me. Supporting Ordered_ste_lang is now even more important as it's going to be a mechanism for controlling bulk builds in conjunction with `(env ..)`

Reply:
master now has `Ordered_set_lang.Unexpanded.fold_strings` and `String_with_vars.text_only` which should allow implementing this.